[PATCH] splice: missing user pointer access verification (CVE-2008-0009/10)

From: Jens Axboe <jens.axboe[at]oracle.com>

vmsplice_to_user() must always check the user pointer and length
with access_ok() before copying. Likewise, for the slow path of
copy_from_user_mmap_sem() we need to check that we may read from
the user region.

Signed-off-by: Jens Axboe <jens.axboe[at]oracle.com>
Cc: Wojciech Purczynski <cliph[at]research.coseinc.com>
Signed-off-by: Greg Kroah-Hartman <gregkh[at]suse.de>


This patch also includes a fix to get_iovec_page_array(), submitted
to the LKML by Bastian Blank <bastian[at]waldi.eu.org>


--- linux-2.6.23.8/fs/splice.c.orig	2008-03-28 22:12:10.000000000 -0700
+++ linux-2.6.23.8/fs/splice.c	2008-03-28 22:22:28.000000000 -0700
@@ -1234,6 +1234,9 @@ static int copy_from_user_mmap_sem(void 
 {
 	int partial;
 
+	if (!access_ok(VERIFY_READ, src, n))
+		return -EFAULT;
+
 	pagefault_disable();
 	partial = __copy_from_user_inatomic(dst, src, n);
 	pagefault_enable();
@@ -1288,6 +1291,10 @@ static int get_iovec_page_array(const st
 		error = -EFAULT;
 		if (unlikely(!base))
 			break;
+		if (unlikely(!access_ok(VERIFY_READ, base, len))) {
+			error = -EFAULT;
+			break;
+		}
 
 		/*
 		 * Get this base offset and number of pages, then map
@@ -1441,6 +1448,10 @@ static long vmsplice_to_user(struct file
 			error = -EFAULT;
 			break;
 		}
+		if (unlikely(!access_ok(VERIFY_WRITE, base, len))) {
+			error = -EFAULT;
+			break;
+		}
 
 		sd.len = 0;
 		sd.total_len = len;
